server/src/main/java/com/saralein/server/HttpServer.java

```diff
@@ -1,45 +1,85 @@
 import com.saralein.server.router.Router;
 import com.saralein.server.setup.Server;
 import com.saralein.server.setup.SetupServer;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
 
 public class HttpServer {
     private final Logger logger;
     private Routes routes;
+    private Router router;
+    private List<Middleware> middlewareList = new ArrayList<>();
+    private Path root = null;
 
     public HttpServer(Logger logger) {
         this.logger = logger;
         this.routes = new Routes();
+        this.router = new Router(routes);
     }
 
-    public HttpServer config(Routes routes) {
-        this.routes = routes;
+    public HttpServer addStatic(Path root) {
+        this.root = root;
         return this;
     }
 
-    public void start(int port, Path root) {
-        Runtime runtime = Runtime.getRuntime();
-        FileHelper fileHelper = new FileHelper(root);
-
-        DirectoryController directoryController = new DirectoryController(fileHelper);
-        FileController fileController = new FileController(fileHelper);
+    public HttpServer router(Routes routes) {
+        this.routes = routes;
+        this.router = new Router(routes);
+        return this;
+    }
 
-        ErrorController errorController = new ErrorController();
+    public HttpServer use(Middleware middleware) {
+        middlewareList.add(middleware);
+        return this;
+    }
 
-        Router router = new Router(directoryController, fileController, errorController, routes, root);
+    public void start(int port) {
+        Runtime runtime = Runtime.getRuntime();
+        Controller application = configureMiddleware();
         RequestParser requestParser = new RequestParser();
         ResponseSerializer responseSerializer = new ResponseSerializer();
-        Server server = new SetupServer(logger, runtime, router, requestParser, responseSerializer).setup(port);
+        Server server = new SetupServer(logger,runtime, application, requestParser, responseSerializer).setup(port);
 
         server.run();
     }
+
+    private Controller configureMiddleware() {
+        return addClientMiddleware(addStaticMiddlewareIfApplicable(router));
+    }
+
+    private Controller addStaticMiddlewareIfApplicable(Controller router) {
+        if (this.root != null) {
+            FileHelper fileHelper = new FileHelper(root);
+
+            DirectoryController directoryController = new DirectoryController(fileHelper);
+            FileController fileController = new FileController(fileHelper);
+            Middleware staticMiddleware = new StaticMiddleware(root, directoryController, fileController);
+            return staticMiddleware.use(router);
+        } else {
+            return router;
+        }
+    }
+
+    private Controller addClientMiddleware(Controller application) {
+        Controller modifiedApplication = application;
+
+        for(int i = middlewareList.size() - 1; i >= 0; i--) {
```

Comment:
You can use `middlewareList.forEach` here.

```java
middlewareList.forEach(middleware -> middleware.use(modifiedApplication));
```

Reply:
Also, the type of the `middlewareList` field is `List<Middleware>`. Maybe we could rename that `middleware` or `middlewares`? 

In general I try to avoid including the type in the variable name. That information is already included in type, I and I don't want to make Java more verbose than it needs to be 😄